cmd/segment-reaper/observer.go

```diff
@@ -143,11 +168,139 @@ func (obsvr *observer) processSegment(ctx context.Context, path metainfo.ScopedP
 // analyzeProject analyzes the objects in obsv.objects field for detecting bad
 // segments and writing them to objs.writer.
 func (obsvr *observer) analyzeProject(ctx context.Context) error {
-	// TODO this part will be implemented in next PR
-	// TODO(if): For what is this?
+	for bucket, objects := range obsvr.objects {
+		for path, object := range objects {
+			if object.skip {
+				continue
+			}
+
+			brokenObject := false
+			if !object.hasLastSegment {
+				brokenObject = true
+			} else {
+				includeLastSegment := true
+				segmentsCount := object.segments.Count()
+
+				// using 'expectedNumberOfSegments-1' because 'segments' doesn't contain last segment
+				switch {
+				// this case is only for old style pointers with encrypted number of segments
+				// value 0 means that we don't know how much segments object should have
+				case object.expectedNumberOfSegments == 0:
+					// verify if initial sequence is valid
+					lastSequenceIndex := 0
+					for ; lastSequenceIndex < maxNumOfSegments; lastSequenceIndex++ {
+						has, err := object.segments.Has(lastSequenceIndex)
+						if err != nil {
+							panic(err)
+						}
+						if !has {
+							break
+						}
+					}
+					// unset valid sequence and mark as broken but don't remove last segment
+					for index := 0; index < lastSequenceIndex; index++ {
+						err := object.segments.Unset(index)
+						if err != nil {
+							panic(err)
+						}
+					}
+					// in this case we always mark object as broken because if
+					// there was valid sequence we are removing it from bitmask and leave only
+					// zombie segments
+					brokenObject = true
```

Comment:
Shouldn't we check if there are any 1s left in the bitmask before setting the object as broken? Otherwise, I don't see the case when an old-style pointer is good.

Reply:
I thought about that but I understood that the result will be the same, although there will be more iterations.

If I'm not mistaken, the segments will be iterated, but because all the bits would be 0, it won't print any segment.